### PR TITLE
Fix some bugs and make core package richer

### PR DESCRIPTION
**Pull request by MarcelGarus**
This PR adds some features to the core package. It also fixes bugs that were discovered while doing this and while fuzzing the compiler.

### CODE REVIEW

File: compiler/frontend/src/hir_to_mir.rs [hunk 1]
```diff
@@ -267,27 +275,25 @@ impl<'a> LoweringContext<'a> {
...
-                    if !pattern.contains_captured_identifiers() {
-                        body.push_reference(nothing)
-                    } else {
-                        // The generated expression will be followed by at least one
-                        // identifier reference, so we don't have to generate a
-                        // reference to the result in here.
-                        self.last_pattern_result = Some((pattern_result, false));
-                        pattern_result
-                    }
+                    let pattern_result = body.push_reference(pattern_result);
```
Comment (JonasWanke) [inline compiler/frontend/src/hir_to_mir.rs]: Do we actually have to push this reference? Subsequent `PatternIdentifierReference`s will refer to the correct ID using `self.ongoing_destructuring` and a `Nothing` will follow so it also can't be the surrounding body's return value

File: packages/Core/controlFlow.candy [hunk 2]
```diff
@@ -32,3 +34,14 @@ loop body :=
       "The `body` has to return `Continue` or `Break`."
     if (result | equals Continue) { recurse Nothing }
   }
+
+repeat times body :=
+  needs (typeOf times | equals Int)
+  needs (times | ✨.intCompareTo 0 | equals Less | bool.not)
+  needs (function.is0 body)
```
Comment (JonasWanke) [inline packages/Core/controlFlow.candy:41]: In Kotlin, the `body` accepts the repetition index as a parameter. We could even support `function.is0` _and_ `function.is1` here, but could also say that if you need the index, you should create a range (once we have that)

File: packages/Core/mutable.candy [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        requests.receivePort | channel.receive %
+          (Get, answerPort) ->
+            needs (channel.isSendPort answerPort)
+            answerPort | channel.send value
+            recurse value
+          (Set, newValue) -> recurse newValue
+          Quit -> value
+      }
+    }
+    await finalValue
```
Comment (JonasWanke) [inline packages/Core/mutable.candy]: Couldn't we remove the `async`/`await` around here and call `recursive` directly?

**Review by JonasWanke — CHANGES_REQUESTED** [review]

File: packages/Core/controlFlow.candy (hunk 2 above)
Comment (MarcelGarus) [reply packages/Core/controlFlow.candy:41]: I like the option of explicitly specifying a range.

File: compiler/frontend/src/hir_to_mir.rs (hunk 1 above)
Comment (MarcelGarus) [reply compiler/frontend/src/hir_to_mir.rs]: I simplified the code a bit by just returning a nothing. We could further simplify the HIR by just not pushing anything at all for this expression, but it makes the rest of this function more complicated (we'd need the expression compilation to produce an `Option<Id>` everywhere).

Considering that tree shaking is a thing, I'm fine with the current solution.

File: compiler/frontend/src/hir_to_mir.rs [hunk 4]
```diff
@@ -268,27 +276,24 @@ impl<'a> LoweringContext<'a> {
...
-                        // identifier reference, so we don't have to generate a
-                        // reference to the result in here.
-                        self.last_pattern_result = Some((pattern_result, false));
-                        pattern_result
-                    }
+                    self.ongoing_destructuring = Some(OngoingDestructuring {
+                        result: pattern_result,
+                        is_trivial: false,
+                    });
+                    body.push_nothing()
```
Comment (JonasWanke) [inline compiler/frontend/src/hir_to_mir.rs]: By moving the `self.ongoing_destructuring = …` above `body.push_if_else(…)`, we can get rid of the `body.push_nothing()`

**Review by JonasWanke — CHANGES_REQUESTED** [review]

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**